pil-analyzer/src/type_inference.rs

```diff
@@ -0,0 +1,675 @@
+            })?;
+
+        for (arg, param) in arguments.into_iter().zip(params) {
+            self.expect_type(&param, arg)?;
+        }
+        Ok(result_type)
+    }
+
+    fn expect_type<T: FieldElement>(
+        &mut self,
+        expected_type: &Type,
+        expr: &mut Expression<T>,
+    ) -> Result<(), String> {
+        // For literals, we try to store the type here already.
+        // This avoids creating tons of type variables for large arrays.
+        if let Expression::Number(_, annotated_type @ None) = expr {
+            match expected_type {
+                Type::Int => *annotated_type = Some(TypeName::Int),
+                Type::Fe => *annotated_type = Some(TypeName::Fe),
+                Type::Expr => *annotated_type = Some(TypeName::Expr),
+                Type::TypeVar(tv) => *annotated_type = Some(TypeName::TypeVar(tv.clone())),
+                _ => {}
+            };
+        }
+        let inferred_type = self.process_expression(expr)?;
+        self.unifier
+            .unify_types(inferred_type.clone(), expected_type.clone())
+            .map_err(|err| {
+                format!(
+                    "Error checking sub-expression {expr}:\nExpected type: {}\nInferred type: {}\n{err}",
+                    self.substitute_to(expected_type.clone()),
+                    self.substitute_to(inferred_type)
+                )
+            })
+    }
+
+    fn substitute_to(&self, mut ty: Type) -> Type {
+        self.substitute(&mut ty);
+        ty
+    }
+
+    fn substitute(&self, ty: &mut Type) {
+        ty.substitute_type_vars(self.unifier.substitutions());
+    }
+
+    /// Instantiates a type scheme by creating new type variables for the quantified
+    /// type variables in the scheme and adds the required trait bounds for the
+    /// new type variables.
+    /// Returns the new type and a vector of the type variables used for those
+    /// declared in the scheme.
+    fn instantiate_scheme(&mut self, scheme: TypeScheme) -> (Type, Vec<Type>) {
+        let mut ty = scheme.ty;
+        let vars = scheme
+            .vars
+            .bounds()
+            .map(|(_, bounds)| {
+                let new_var = self.new_type_var();
+                for b in bounds {
+                    self.unifier.ensure_bound(&new_var, b.clone()).unwrap();
+                }
+                new_var
+            })
+            .collect::<Vec<_>>();
+        let substitutions = scheme.vars.vars().cloned().zip(vars.clone()).collect();
+        ty.substitute_type_vars(&substitutions);
+        (ty, vars)
+    }
+
+    fn new_type_var_name(&mut self) -> String {
+        self.last_type_var += 1;
+        format!("T{}", self.last_type_var)
+    }
+
+    fn new_type_var(&mut self) -> Type {
+        Type::TypeVar(self.new_type_var_name())
+    }
+
+    /// Creates a type scheme out of a type by making all unsubstituted
+    /// type variables generic.
+    /// TODO this is wrong for mutually recursive generic functions.
```

Comment:
does this mean that for now those would fail type check?

Reply:
I actually haven't tried what happens...
Maybe it works because we specify the type schemes explicitly.